### ISSUE TITLE
y_true gets casted to y_pred.dtype in every losses

### ISSUE THREAD

**Issue by old-school-kid** [keras-team/keras#15014]
Every loss function casts y_true into y_pred data type resulting in error. If y_true>2*32 and y_pred is of data type int32 then an erroneous loss is returned by mean squared error. Similar errors can be found from mean absolute error, categorical cross entropy and sparse categorical cross entropy. 
errors can be found in this [sample notebook](https://colab.research.google.com/drive/1xdWCu4MavfmfimQQlXciPl_Atbip8fWi?usp=sharing)

**Comment by old-school-kid** [comment]
I created a [Pull Request](https://github.com/keras-team/keras/pull/15015#issue-698536884) which tries to solve the problem.

**Comment by rmothukuru** [comment]
@old-school-kid,
Thank you for the issue and for the PR. Our Team will review the PR and will provide the comments accordingly.

**Comment by qlzh727** [comment]
I think the cast is mainly trying to address the dtype mismatch, eg int and float for label and prediction, which is quite common in normal case. The 2*32 seems to be a very extreme case, which we barely hit. Can we change the PR to only cast the dtype if the label and prediction are in different category?

**Comment by zuyezheng** [comment]
This issue can actually surface earlier with mixed precision and float16 much earlier with around 2k categories. With bfloat16 it happens at around 200. Filed this [bug](https://github.com/keras-team/keras/issues/15012) and was asked to dupe it here, I believe @old-school-kid copied the examples from my [notebook](https://colab.research.google.com/drive/1oRbNOnCo1i2HcXD2V4_-D1Bz2EVxiT65) specifically for float16/

Ran into this with huggingface as they don't currently explicitly cast.

**Comment by qlzh727** [comment]
Thanks @zuyezheng for more context, float16 and bfloat16 does seems to be a more common issue here since the overflow limit is quite low here. Let me check the PR. This is likely a breaking change and would raise issue on user side, which we might need to check with API owner.

**Comment by qlzh727** [comment]
Please see my review comments on the PR #15015.
